src/goto-instrument/contracts/contracts.cpp

```diff
@@ -601,22 +656,54 @@ bool code_contractst::apply_function_contract(
 }
 
 void code_contractst::apply_loop_contract(
-  const irep_idt &function,
+  const irep_idt &function_name,
   goto_functionst::goto_functiont &goto_function)
 {
   local_may_aliast local_may_alias(goto_function);
   natural_loops_mutablet natural_loops(goto_function.body);
 
-  // Iterate over the (natural) loops in the function,
-  // and apply any invariant annotations that we find.
+  struct loop_graph_nodet : public graph_nodet<empty_edget>
+  {
+    typedef const goto_programt::targett &targett;
+    typedef const typename natural_loops_mutablet::loopt &loopt;
+
+    targett target;
+    loopt loop;
+
+    loop_graph_nodet(targett t, loopt l) : target(t), loop(l)
+    {
+    }
+  };
+  grapht<loop_graph_nodet> loop_nesting_graph;
+
   for(const auto &loop : natural_loops.loop_map)
+    loop_nesting_graph.add_node(loop.first, loop.second);
+
+  for(size_t outer = 0; outer < loop_nesting_graph.size(); ++outer)
+  {
+    for(size_t inner = 0; inner < loop_nesting_graph.size(); ++inner)
+    {
+      if(inner == outer)
+        continue;
+
+      if(loop_nesting_graph[outer].loop.contains(
+           loop_nesting_graph[inner].target))
+        loop_nesting_graph.add_edge(outer, inner);
+    }
+  }
```

Comment:
This code would really benefit from some comments. Quadratic loops are always scary. Also, I'd at least rewrite this for the inner loop to start from `outer + 1` and test both inner-containers-outer and outer-contains-inner in the loop body. But that might be a matter of personal preference.

Reply:
I'm not sure ... I think it would lead to mode code duplication (of the containment-checking + edge-adding part):

```c
  for(size_t a = 0; a < loop_nesting_graph.size(); ++a)
  {
    for(size_t b = a + 1; b < loop_nesting_graph.size(); ++b)
    {
      if(loop_nesting_graph[a].loop.contains(
           loop_nesting_graph[b].target))
        loop_nesting_graph.add_edge(a, b);
      else if(loop_nesting_graph[b].loop.contains(
           loop_nesting_graph[a].target))
        loop_nesting_graph.add_edge(b, a);
    }
  }
```